compiler/vmprofiler.nim

```diff
@@ -0,0 +1,41 @@
+
+import
+  options, vmdef, times, std/private/miscdollars, lineinfos, strutils, tables,
+  msgs
+  
+proc enter*(prof: var Profiler, c: PCtx, tos: PStackFrame) {.inline.} =
+  if optProfileVM in c.config.globalOptions:
+    prof.tEnter = cpuTime()
+    prof.tos = tos
+
+proc leave*(prof: var Profiler, c: PCtx) {.inline.} =
+  if optProfileVM in c.config.globalOptions:
+    let tLeave = cpuTime()
+    var tos = prof.tos
+    while tos != nil:
+      if tos.prc != nil:
+        let li = tos.prc.info
+        if li notin c.profiler.data:
+          c.profiler.data[li] = ProfileInfo()
+        c.profiler.data[li].time += tLeave - prof.tEnter
+        inc c.profiler.data[li].count
+      tos = tos.next
+
+proc dump*(p: var Profiler, c: PCtx) =
+  if optProfileVM in c.config.globalOptions:
+    echo "\nprof:     µs     count  location"
+    var data = c.profiler.data
+    for i in 0..<32:
+      var tMax: float
+      var infoMax: ProfileInfo
+      var flMax: TLineInfo
+      for fl, info in data:
+        if info.time > infoMax.time:
+          infoMax = info
+          flMax = fl
+      if infoMax.count == 0:
+        break
+      echo  "  " & align($int(infoMax.time * 1e6), 10) &
```

Comment:
when I run your code the location info points to the location where symbol is defined, not to the location where line is executed (as happens when you show a nim stacktrace), which is less informative

Reply:
True, because I use the top stack frame for profiling - I personally prefer my profiles at the function granularity, since that makes sense if you make the profile fully inclusive/accumulative.

Reply:
Too late, already answered! :)